File: javascript/functional-js/pipe.js
// pipe(value, ...function) -> value
// Transforms a value through a series of single-argument functions.
// pipe("  hEllO, worLd   ", (text) => text.toLowerCase(), (text) => text.trim()) -> "hello, world"

function pipe(input, ...actions) {
    return actions.reduce(
        (accum, action) => action(accum),
        input
    );
}

// compose(...function) -> function -> value
// Composes a series of single-argument functions into a single function expression.
// const edit = compose((text) => text.toLowerCase(), (text) => text.trim()) -> edit("  hEllO, worLd   ") -> "hello, world"

function compose(...actions) {
    return function (input) {
        return actions.reduce(
            (accum, action) => action(accum), 
            input
        );
    };
}

File: python/functional-py/pipe.py
# compose(*functions) -> function
# Composes a series of single-argument functions into a one function.
# sums = compose(lambda x: x + 1, lambda x: x * 10)
# sums(6) -> 70

def compose(*actions):
    def reducer(input):
        accum = input
        for action in actions:
            accum = action(accum)
        return accum
    return reducer
 
# pipe(value, *function) -> value
# Transforms a value through a series of single-argument functions.
# pipe(6, lambda x: x + 1, lambda x: x * 10) -> 70

def pipe(input, *actions):
    accum = input
    for action in actions:
        accum = action(accum)
    return accum
